clang/lib/CodeGen/ItaniumCXXABI.cpp

```diff
@@ -2794,6 +2794,14 @@ void ItaniumCXXABI::registerGlobalDtor(CodeGenFunction &CGF, const VarDecl &D,
   if (D.isNoDestroy(CGM.getContext()))
     return;
 
+  // OpenMP offloading supports C++ constructors and destructors but we do not
+  // always have 'atexit' available. Instead lower these to use the LLVM global
+  // destructors which we can handle directly in the runtime.
+  if (CGM.getLangOpts().OpenMP && CGM.getLangOpts().OpenMPIsTargetDevice &&
+      !D.isStaticLocal() &&
+      (CGM.getTriple().isAMDGPU() || CGM.getTriple().isNVPTX()))
```

Comment:
Oh look, it's both of my favorite patterns. Can you refine this into something better than language X | language Y and AMDGPU || PTX

Reply:
Yeah, these types of things are problematic especially if we consider getting SPIR-V support eventually. The logic basically goes like this. OpenMP supports global destructors but does not always support the `atexit` function. The old logic used to replace everything. This now at least lets CPU based targets use regular handling. I could make this unconditional for OpenMP, but I figured it'd be better to allow the CPU based targets to use the regular handling.

More or less this is just a concession to prevent regressions from this patch. The old logic looked like this, which did this unconditionally. Like I said, could remove the AMD and PTX checks and just do this on the CPU as well if it would be better.
```c++
  if (CGM.getLangOpts().OMPTargetTriples.empty() &&
      !CGM.getLangOpts().OpenMPIsTargetDevice)
    return false;
```

Reply:
Just make this apply to all triples. I don't want to remove the dependency on the OpenMP language because this is somewhat of a hack. We can revisit this later if needed.

Reply:
Would also just hide this in a target/lang predicate that lists these

Reply:
So just some random helper function like "Does target support X?"

Reply:
I could put something in `LangOptions` that just returns the same thing. Wasn't sure if it's worth forcing a recompile of everything though.